File: spectral_library/lib/Models/spect_file.dart
import 'package:spectral_library/Models/category.dart';
import 'package:spectral_library/Models/data.dart';

class SpectFile {
  int? fileId;
  String filename;
  Category category;
  String? description;
  bool isPublic;
  List<Data> dataPoints;
  List<String>? sharedWith;

  SpectFile({
    this.fileId,
    required this.filename,
    required this.category,
    required this.isPublic,
    required this.dataPoints,
    this.description,
    this.sharedWith,
  });

  /// Convert a SpectFile object to a Map for serialization
  Map<String, dynamic> toMap() {
    return {
      'fileId': fileId,
      'filename': filename,
      // Use category.toMap() instead of just category
      'category': category.toMap(),
      'description': description,
      'isPublic': isPublic,
      'dataPoints': dataPoints.map((data) => data.toMap()).toList(),
      'sharedWith': sharedWith,
    };
  }

  /// Create a SpectFile object from a Map
  factory SpectFile.fromMap(Map<String, dynamic> map) {
    return SpectFile(
      fileId: map['fileId'] as int?,
      filename: map['filename'] as String,
      // Convert the category map into a Category object
      category: Category.fromMap(map['category'] as Map<String, dynamic>),
      description: map['description'] as String?,
      isPublic: map['isPublic'] as bool,
      dataPoints: (map['dataPoints'] as List<dynamic>)
          .map((dataMap) => Data.fromMap(dataMap as Map<String, dynamic>))
          .toList(),
      sharedWith: (map['sharedWith'] as List<dynamic>?)
          ?.map((item) => item as String)
          .toList(),
    );
  }
}

File: backend/Models/SpectFile.py
from typing import Optional, List

from Models.Data import Data

class SpectFile:
    def __init__(
        self,
        filename: str,
        category: str,
        is_public: bool,
        data_points: List[Data],
        file_id: Optional[int] = None,
        description: Optional[str] = None,
        shared_with: Optional[List[str]] = None,
    ):
        self.file_id = file_id
        self.filename = filename
        self.category = category
        self.description = description
        self.is_public = is_public
        self.data_points = data_points
        self.shared_with = shared_with or []

    def to_map(self):
        return {
            "fileId": self.file_id,
            "filename": self.filename,
            "category": self.category,
            "description": self.description,
            "isPublic": self.is_public,
            "dataPoints": [data.to_map() for data in self.data_points],
            "sharedWith": self.shared_with,
        }
    @classmethod
    def from_map(cls, data):
        return cls(
            file_id=data.get("fileId"),
            filename=data["filename"],
            category=data["category"],
            description=data.get("description"),
            is_public=data["isPublic"],
            data_points=[Data.from_map(dp) for dp in data["dataPoints"]],
            shared_with=data.get("sharedWith", []),
        )
